### ISSUE TITLE
minimal build gives severely crippled text selection behaviour

### ISSUE THREAD

**Issue by RJVB** [annulen/webkit#688]
I've been experimenting with a full vs. a minimal packaging build, as discussed with @annulen. The minimal build mostly works just as the full build as judged from using Qt's Assistant for documentation browsing but I just discovered that it has severely crippled text selection support.

My current minimal build config is

```
-DENABLE_API_TESTS=OFF 
-DENABLE_PRINT_SUPPORT=ON
-DENABLE_DEVICE_ORIENTATION=OFF
-DENABLE_GEOLOCATION=OFF
-DENABLE_JIT=ON
-DENABLE_OPENGL=ON
-DENABLE_WEBKIT2=OFF
-DENABLE_QT_WEBCHANNEL=OFF
-DUSE_LIBHYPHEN=ON
-DUSE_GSTREAMER=OFF
-DENABLE_INSPECTOR_UI=OFF
-DENABLE_DRAG_SUPPORT=OFF
-DENABLE_SAMPLING_PROFILER=OFF
-DENABLE_VIDEO=OFF
-DENABLE_VIDEO_TRACK=OFF
```

The only difference that seems relevant at first glance is drag support; I'll reactivate it and report back.

**Comment by annulen** [comment]
`-DENABLE_DRAG_SUPPORT=OFF` is doing this. Please don't disable features if you don't know their meaning, especially private features

**Comment by annulen** [comment]
See also https://github.com/annulen/webkit/blob/qtwebkit-5.212/Source/cmake/OptionsQt.cmake#L251-L253

**Comment by RJVB** [comment]
Yeah, but the drag option is defined above that warning, along with a couple of others you suggested yourself I might turn off ;)
I can't remember if I discovered the drag support option myself or not, but it apparently doesn't control drag/drop support alone but also support for actions typically used before you start dragging.

If normal, full-featured selection cannot be separated from drag/drop support it might not hurt to rename the option to make that clear (or at least add a comment).

**Comment by annulen** [comment]
My bad, it should be private probably

>If normal, full-featured selection cannot be separated from drag/drop support it might not hurt to rename the option to make that clear (or at least add a comment).

Possible

**Comment by annulen** [comment]
https://bugs.webkit.org/show_bug.cgi?id=180266

**Comment by RJVB** [comment]
++

**Comment by annulen** [comment]
About other options:
* disabling inspector UI is a dubious thing - on one side it reduces size, however when you then need inspector it's hard to make it wroking. It may be a better idea to build it as a binary resource file instead
* video track is automatically off when you disable video

**Comment by RJVB** [comment]
>* disabling inspector UI is a dubious thing - on one side it reduces size, however when you then need inspector it's hard to make it wroking.

That's something you can say about all choices I made here. "When you need video it's hard to make it work" when video support is disabled. 

Video support probably takes much more size and resources and can probably not be built any other way; I picked the example for comparison because you could actually need video support for some kind of more advanced documentation (tutorials) whereas the inspector is a feature mostly aimed at web (or webkit) developers IMHO.

About setting ENABLE_DRAG_SUPPORT back to ON: why is it that this caused a full rebuild? Does the current cmake build system set all preprocessor macros at the toplevel, so that all source has to be rebuilt even if only a single file actually checks a given macro?

**Comment by annulen** [comment]
>"When you need video it's hard to make it work" when video support is disabled.

But video support enables a lot of extra code and pulls in dependencies (ok, not so much on macOS and it's probably possible to get rid of QtMultimedia by using AVFoundation player), while disabling inspector UI removes no native code, just some resource files.

>inspector is a feature mostly aimed at web (or webkit) developers IMHO.

It helps a lot when QtWebKit misbehaves for some user or some application and bug is not easy to reproduce. It also has more mundane uses e.g. downloading resources which page author didn't want to be downloadable, removing cruft from page before printing when there is no proper printable version, etc.

>why is it that this caused a full rebuild

Any option declared with `WEBKIT_OPTION_...` causes full rebuild when changed

> Does the current cmake build system set all preprocessor macros at the toplevel, so that all source has to be rebuilt even if only a single file actually checks a given macro?

Yes, though many features are quite invasive, e.g. for DRAG_SUPPORT git grep reports 80 files are affected.

**Comment by annulen** [comment]
Fixed in trunk, cherry-picked as e0cac3c0f63ad4db520794180f388bde22f04f53
